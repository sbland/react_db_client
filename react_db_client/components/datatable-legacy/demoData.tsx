/* eslint-disable react/prop-types */
import React, { useEffect, useMemo } from 'react';
import {
  EComparisons,
  EFilterType,
  FilterObjectClass,
  FilterOption,
  Uid,
} from '@react_db_client/constants.client-types';
import { IHeadingCustomExample, THeading } from './lib';

export const demoFilterString = new FilterObjectClass({
  uid: 'demoFilterString',
  field: 'name',
  value: 'Foo',
  label: 'Demo Filter String',
  operator: EComparisons.CONTAINS,
  type: EFilterType.text,
});
export const demoFilterNumber = new FilterObjectClass({
  uid: 'demoFilterNumber',
  field: 'count',
  label: 'Demo Filter Number',
  value: 0,
  // step: 1,
  operator: EComparisons.GREATER_THAN,
  type: EFilterType.number,
});
// export const demoFilterExpression = new FilterObjectClass({
//   uid: 'demoFilterExpression',
//   field: 'expression',
//   label: 'Demo Filter Expression',
//   value: 0,
//   operator: '>',
// });

export const demoFilterSelect = new FilterObjectClass({
  uid: 'demoFilterExpression',
  field: 'select',
  label: 'Demo Filter Select',
  value: 'a',
  operator: EComparisons.CONTAINS,
  type: EFilterType.select,
});

export const demoFiltersData = [demoFilterString, demoFilterNumber];

export const demoTableDataSimple = [
  {
    uid: 'a',
    name: 'Foo',
    count: 4,
  },
  {
    uid: 'b',
    name: 'Bar',
    count: 3,
  },
  {
    uid: 'c',
    name: 'Roo',
    count: -100,
  },
  {
    uid: 'd',
    name: 'Foo',
    count: 2,
  },
  {
    uid: 'e',
    name: '',
    count: 2,
  },
];

// The demo data after filter, sort and evaluate
export const demoTableDataEvaluatedSimple = [
  {
    uid: 'd',
    name: 'Foo',
    count: 2,
    eval: 3,
  },
  {
    uid: 'a',
    name: 'Foo',
    count: 4,
    eval: 5,
  },
];

export const demoHeadingsDataSimple = [
  {
    uid: 'uid',
    label: 'UID',
    type: 'button',
    action: (x) => alert(x),
  },
  {
    uid: 'name',
    label: 'Name',
    type: EFilterType.text,
  },
  {
    uid: 'count',
    label: 'Count',
    type: EFilterType.number,
    defaultValue: 7,
    step: 1,
  },
  {
    uid: 'eval',
    label: 'Eval',
    evaluateType: EFilterType.number,
    type: EFilterType.number,
    expression: '$count + 1',
    expressionReversed: '$count=$_ - 1',
  },
  {
    uid: 'select',
    label: 'Select',
    type: 'select',
    options: [
      { uid: 'a', label: 'A' },
      { uid: 'b', label: 'B' },
      { uid: 'c', label: 'C' },
    ],
  },
  {
    uid: 'hidden',
    label: 'Hidden',
    type: 'text',
    hidden: true,
  },
];

export const demoHeadingsSimpleHiddenIds = demoHeadingsDataSimple
  .filter((h) => h.hidden)
  .map((h) => h.uid);

export const demoDataSimpleTotals = {
  count: 7,
};

export const demoTableData = [
  {
    uid: 'a',
    natid: '100a',
    name: 'Foo',
    count: 3,
    multiplier: 2,
    toggle: true,
    readOnly: "Can't touch me",
    select: 'a',
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
  {
    // Note string "4" here to check we can pass string numbers without issues
    uid: 'ab',
    natid: '10a',
    name: 'Foobar',
    count: '99',
    def: 3,
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
  {
    uid: 'b',
    natid: '50a',
    name: 'Bar',
    description: 'A really really really long description that needs to fit into a little box!',
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
  {
    uid: 'c',
    name: 'C',
    count: 3,
    multiplier: 3,
    expression: 9,
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
  {
    uid: 'd',
    name: '',
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
];

export const demoTotals = {
  uid: 'na',
  name: 'na',
  count: 10,
  multiplier: 10,
  toggle: 'na',
  readOnly: 'na',
  select: 'na',
  hiddenDemoNumber: 8,
  hiddenDemo: 'na',
};

export const demoTableDataLong = [...Array(100)]
  .map((v, i) => ({
    uid: i,
    name: 'Foo',
    count: 3,
    multiplier: 2,
    toggle: true,
    readOnly: "Can't touch me",
    select: 'a',
  }))
  .reduce((acc, v) => {
    acc[v.uid] = v;
    return acc;
  }, {});

// export const demoTableData = Object.values(demoTableDataObject);

export const demoTableDataEvaluated = [
  {
    count: 3,
    // def: 3,
    hiddenDemoNumber: 3,
    expression: 6,
    multiplier: 2,
    hiddenDemo: 'Hide me',
    name: 'Foo',
    toggle: true,
    uid: 'a',
    validated: false,
    invalid: 'INVALID',
    invalidb: 'INVALID',
    readOnly: "Can't touch me",
    select: 'a',
  },
  {
    count: 4,
    hiddenDemoNumber: 3,
    def: 3,
    hiddenDemo: 'Hide me',
    expression: 'INVALID',
    invalid: 'INVALID',
    invalidb: 'INVALID',
    name: 'Foobar',
    uid: 'ab',
    validated: 'INVALID',
  },
];

export const demoTableTotals = {
  count: 7,
  def: 3,
  expression: 6,
  hiddenDemoNumber: 6,
  multiplier: 2,
  invalid: 0,
  invalidb: 0,
};

export const demoProcessedData = {
  tableDataFiltered: demoTableDataEvaluated,
  totals: demoTableTotals,
};

export const demoHeadingsData: THeading<IHeadingCustomExample>[] = [
  {
    uid: 'uid',
    label: 'UID',
    type: EFilterType.button,
    action: (x) => alert(x),
    styleRule: '$count<5',
    unique: true,
    required: true,
  },
  {
    uid: 'name',
    label: 'Name',
    type: EFilterType.text,
    unique: true,
    required: true,
  },
  {
    uid: 'description',
    label: 'description',
    type: EFilterType.textLong,
  },
  {
    uid: 'count',
    label: 'Count',
    type: EFilterType.number,
    max: 10,
    min: 2,
    defaultValue: 8,
    step: 1,
    showTotals: true,
  },
  {
    uid: 'def',
    label: 'Def',
    type: EFilterType.number,
    max: 10,
    min: 1,
    defaultValue: 7,
    showTotals: true,
  },
  {
    uid: 'multiplier',
    label: 'Multiplier',
    type: EFilterType.number,
    max: 10,
    min: 1,
    // defaultValue: '$def', // NOT IMPLEMENTED
  },
  {
    uid: 'readOnly',
    label: 'Read Only',
    type: EFilterType.text,
    readOnly: true,
  },
  {
    uid: 'hiddenDemo',
    label: 'Hidden',
    type: EFilterType.text,
    hidden: true,
  },
  {
    uid: 'hiddenDemoNumber',
    label: 'Hidden Number',
    type: EFilterType.number,
    hidden: true,
  },
  {
    uid: 'toggle',
    label: 'Toggle',
    type: EFilterType.toggle,
  },
  {
    uid: 'toggleR',
    label: 'ToggleR',
    type: EFilterType.toggle,
    readOnly: true,
  },
  {
    uid: 'expression',
    label: 'Expression',
    type: EFilterType.number,
    evaluateType: EFilterType.number,
    expression: '$count * $multiplier',
    expressionReversed: '$count=$_ / $multiplier',
    // expressionReversed: '$count * $multiplier',
    showTotals: true,
  },
  {
    uid: 'validated',
    label: 'Is Validated',
    evaluateType: EFilterType.text,
    type: EFilterType.bool,
    operator: EComparisons.EQUALS,
    target: 'a',
    field: 'select',
    invert: true,
  },
  {
    uid: 'invalid',
    label: 'Invalid',
    evaluateType: EFilterType.number,
    type: EFilterType.number,
    expression: '$nothing',
    expressionReversed: '$nothing',
  },
  {
    uid: 'invalidb',
    label: 'InvalidB',
    type: EFilterType.number,
    evaluateType: EFilterType.number,
    expression: '$invalid * 2',
    expressionReversed: '$invalid / 2',
  },
  {
    uid: 'select',
    label: 'Select',
    type: EFilterType.select,
    options: [
      { uid: 'a', label: 'A' },
      { uid: 'b', label: 'B' },
      { uid: 'c', label: 'C' },
    ],
  },
  {
    uid: 'internalRoom',
    label: 'Internal Room',
    type: EFilterType.reference,
    collection: 'rooms',
  },
  {
    uid: 'longHeading',
    label: 'Long Heading which is really really long',
    type: EFilterType.text,
  },
  {
    uid: 'customfield',
    label: 'Custom',
    type: 'custom',
    isCustomType: true,
    showTotals: true,
  },
  {
    uid: 'customfieldeval',
    label: 'Custom Eval',
    type: 'customeval',
    isCustomType: true,
    showTotals: true,
    expression: '$customfield + 1',
    expressionReversed: '$customfield=$_ - 1',
    evaluateType: EFilterType.number,
  },
];

export const demoTableDataEvaluationTable = [
  {
    uid: 'a',
    a: 1,
    b: 1,
    c: 1,
    count: 1,
    totalc: 1,
    d: 0,
    lineTotal: 1,
  },
  {
    uid: 'b',
    a: 2,
    b: 2,
    c: 4,
    count: 7,
    totalc: 8,
    d: 33.33,
    lineTotal: 14,
  },
];

export const evaluationTableHeadings = [
  {
    uid: 'uid',
    label: 'Id',
    type: 'text',
  },
  {
    uid: 'a',
    label: 'A',
    type: EFilterType.number,
    // readOnly: true,
    def: 1,
    step: 0.01,
  },
  {
    uid: 'b',
    label: 'B',
    type: EFilterType.number,
    def: '$b-default',
    step: 0.01,
  },
  {
    uid: 'c',
    label: 'C',
    evaluateType: EFilterType.number,
    type: EFilterType.number,
    expression: '$a + $b',
    expressionReversed: '$b=$_ - $a',
    step: 0.01,
  },
  {
    uid: 'count',
    label: 'Count',
    type: EFilterType.number,
    // readOnly: true,
    step: 1,
  },
  {
    uid: 'totalc',
    label: 'Total C',
    type: EFilterType.number,
    evaluateType: EFilterType.number,
    expression: '$count * $c',
    step: 0.01,
    showTotals: true,
    readOnly: true,
  },
];
export const CustomField = ({ acceptValue, cellData, editMode, focused }) => {
  const actionFunc = useMemo(() => {
    return () => acceptValue(Math.random());
  }, [acceptValue]);

  useEffect(() => {
    if (editMode && focused) {
      actionFunc();
    }
  }, [editMode, focused, actionFunc]);
  return (
    <div className="dataTableCellData dataTableCellData-button">
      <button type="button" onClick={actionFunc}>
        {cellData}
      </button>
    </div>
  );
};

// eslint-disable-next-line no-unused-vars
export const customFilter = (value, expression, targetValue) => {
  if (Math.abs(value - targetValue) < 0.3) return true;
  return false;
};

// export const DemoTotalsWrapper = () => {
//   const [total, setTotal] = useState(null);
//   return (
//     <div>
//       <DataTable
//         tableData={demoTableData}
//         headingsData={demoHeadingsData}
//         handleSaveData={(data) => console.log(data)}
//         outputTotals={setTotal}
//         showTotals
//       />
//       <h3>Totals</h3>
//       <p className="total">{JSON.stringify(total)}</p>
//     </div>
//   );
// };

export const availableFilters = demoHeadingsData.reduce(
  (acc, h) => ({ ...acc, [h.uid]: new FilterOption({ ...h, field: h.field || h.uid }) }),
  {} as { [k: Uid]: FilterOption }
);
