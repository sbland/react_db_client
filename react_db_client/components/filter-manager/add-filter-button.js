import React from 'react';
import PropTypes from 'prop-types';


import {
  FilterObjectClass,
} from '@samnbuk/react_db_client.constants.client-types';

export const AddFilterButton = ({
  fieldsData,
  returnNewFilter,
  customFilters,
}) => {
  const handleNewFilter = () => {
    // Add a new filter and use the first field as initial field
    const initialFieldData = Object.values(fieldsData)[0];
    const { uid, field, type, label } = initialFieldData;
    const newFilter = new FilterObjectClass({
      field: field || uid,
      type,
      label,
      filterOptionId: uid,
      isCustomType: Object.keys(customFilters).indexOf(type) >= 0,
    });
    returnNewFilter(newFilter);
  };

  return (
    <button
      type="button"
      className="button-two addFilterBtn"
      onClick={handleNewFilter}
    >
      Add Filter
    </button>
  );
};

AddFilterButton.propTypes = {
  returnNewFilter: PropTypes.func.isRequired,
  fieldsData: PropTypes.objectOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
    })
  ).isRequired,
  customFilters: PropTypes.objectOf(PropTypes.func),
};

AddFilterButton.defaultProps = {
  customFilters: {},
};
