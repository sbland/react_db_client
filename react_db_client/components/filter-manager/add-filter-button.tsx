import React from 'react';
import PropTypes from 'prop-types';

import { FilterObjectClass, FilterOption } from '@react_db_client/constants.client-types';
import { TFilterFunc } from './lib';

export interface IAddFilterButtonProps {
  fieldsData: { [key: string]: FilterOption<any, boolean> };
  returnNewFilter: <Vtype, IsCustomType extends boolean>(
    newFilter: FilterObjectClass<Vtype, IsCustomType>
  ) => void;
  customFilters?: { [key: string]: TFilterFunc };
}

export const AddFilterButton: React.FC<IAddFilterButtonProps> = ({
  fieldsData,
  returnNewFilter,
  customFilters = {},
}) => {
  const handleNewFilter = () => {
    // Add a new filter and use the first field as initial field
    const initialFieldData = Object.values(fieldsData)[0];
    const { uid, field, type, label } = initialFieldData;
    const isCustomFieldType = Object.keys(customFilters).indexOf(type) >= 0;

    const newFilter = new FilterObjectClass({
      field: field || uid,
      type,
      label,
      filterOptionId: uid,
      isCustomType: isCustomFieldType,
    });
    returnNewFilter(newFilter);
  };

  return (
    <button type="button" className="button-two addFilterBtn" onClick={handleNewFilter}>
      Add Filter
    </button>
  );
};

AddFilterButton.propTypes = {
  returnNewFilter: PropTypes.func.isRequired,
  fieldsData: PropTypes.objectOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
    })
  ).isRequired,
  customFilters: PropTypes.objectOf(PropTypes.func),
};

AddFilterButton.defaultProps = {
  customFilters: {},
};
