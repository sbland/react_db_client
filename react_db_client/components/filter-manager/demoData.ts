import {
  comparisons,
  filterTypes,
  FilterObjectClass,
  EFilterType,
} from '@react_db_client/constants.client-types';
import { IField } from './lib';

export const demoFilterString = new FilterObjectClass({
  uid: 'demoFilterString',
  field: 'name',
  value: 'Foo',
  operator: comparisons.contains,
  type: filterTypes.text,
});
export const demoFilterNumber = new FilterObjectClass({
  uid: 'demoFilterNumber',
  field: 'count',
  value: 0,
  operator: comparisons.greaterThan,
  type: filterTypes.number,
});
// export const demoFilterExpression = new FilterObjectClass({
//   uid: 'demoFilterExpression',
//   field: 'expression',
//   value: 0,
//   operator: '>',
// });

export const demoFilterSelect = new FilterObjectClass({
  uid: 'demoFilterExpression',
  field: 'select',
  value: 'a',
  operator: comparisons.contains,
  type: filterTypes.select,
});

export const demoFieldsData = {
  name: {
    uid: 'name',
    label: 'Name',
    type: filterTypes.text,
  },
  count: {
    uid: 'count',
    label: 'Count',
    type: filterTypes.number,
  },
  filterA: {
    uid: 'filterA',
    label: 'filter A',
    type: filterTypes.text,
  },
  filterB: {
    uid: 'filterB',
    label: 'filter B',
    type: filterTypes.text,
  },
  select: {
    uid: 'select',
    label: 'Select',
    type: filterTypes.select,
    options: [
      { uid: 'a', label: 'A' },
      { uid: 'b', label: 'B' },
    ],
  },
  filterEmbedded: {
    uid: 'filterEmbedded',
    label: 'Embedded Filter',
    type: filterTypes.embedded,
    filters: {
      filterC: {
        uid: 'filterC',
        label: 'filter C',
        type: filterTypes.text,
      },
      filterD: {
        uid: 'filterD',
        label: 'filter D',
        type: filterTypes.text,
      },
    },
  },
};

export const demoFiltersData = [
  demoFilterString,
  demoFilterNumber,
  // demoFilterExpression,
  demoFilterSelect,
];

export const exampleOptions = [
  { uid: 'foo', label: 'Foo' },
  { uid: 'bar', label: 'Bar' },
];

const selectionTypes = [EFilterType.select, EFilterType.selectMulti];

export const allTypeFieldsData = Object.keys(filterTypes)
  .map((key) => ({
    uid: key,
    label: key,
    type: key,
    options: selectionTypes.indexOf(key as EFilterType) !== -1 ? exampleOptions : undefined,
  }))
  .reduce((acc, v) => ({ ...acc, [v.uid]: v }), {}) as Record<EFilterType, IField>;

export const allTypeFilters: FilterObjectClass[] = Object.values(allTypeFieldsData).map(
  (field) =>
    new FilterObjectClass({
      ...field,
      field: field.uid,
      value: undefined,
      // expression: null,
      type: field.type,
    })
);
