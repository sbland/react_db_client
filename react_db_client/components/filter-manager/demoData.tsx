import React from 'react';
import {
  comparisons,
  FilterObjectClass,
  EFilterType,
  EComparisons,
  FilterOption,
} from '@react_db_client/constants.client-types';

export const demoFieldName = new FilterOption<any, false>({
  uid: 'name',
  field: 'name',
  label: 'Name',
  type: EFilterType.text,
});

export const demoFieldCount = new FilterOption<any, false>({
  uid: 'count',
  field: 'count',
  label: 'Count',
  type: EFilterType.number,
});

export const demoFieldSelect = new FilterOption<any, false>({
  uid: 'select',
  field: 'select',
  label: 'Select',
  type: EFilterType.select,
  options: [
    { uid: 'a', label: 'A' },
    { uid: 'b', label: 'B' },
  ],
});

export const customField = new FilterOption<any, true>({
  uid: 'customField',
  field: 'customField',
  type: 'customFieldType',
  label: 'Custom field',
  operators: [EComparisons.EQUALS],
  isCustomType: true,
});

export const demoFieldsData: { [k: string]: FilterOption<any, boolean> } = {
  name: demoFieldName,
  count: demoFieldCount,
  filterA: new FilterOption({
    uid: 'filterA',
    field: 'filterA',
    label: 'filter A',
    type: EFilterType.text,
  }),
  filterB: new FilterOption({
    uid: 'filterB',
    field: 'filterB',
    label: 'filter B',
    type: EFilterType.text,
  }),
  select: demoFieldSelect,
  // TODO: Check if we need to support nested types
  // filterEmbedded: {
  //   uid: 'filterEmbedded',
  //   label: 'Embedded Filter',
  //   type: EFilterType.embedded,
  //   filters: {
  //     filterC: new FilterOption<any, false>({
  //       uid: 'filterC',
  //       field: 'filterC',
  //       label: 'filter C',
  //       type: EFilterType.text,
  //     }),
  //     filterD: new FilterOption<any, false>({
  //       uid: 'filterD',
  //       field: 'filterD',
  //       label: 'filter D',
  //       type: EFilterType.text,
  //     }),
  //   },
  // },
  customField,
};
export const exampleOptions = [
  { uid: 'foo', label: 'Foo' },
  { uid: 'bar', label: 'Bar' },
];

const selectionTypes = [EFilterType.select, EFilterType.selectMulti];

export const allTypeFieldsData = Object.keys(EFilterType)
  .map(
    (key) =>
      ({
        uid: key,
        label: key,
        field: key,
        type: key,
        typeArgs: {
          options: selectionTypes.indexOf(key as EFilterType) !== -1 ? exampleOptions : undefined,
        },
      } as FilterOption<any, false>)
  )
  .reduce((acc, v) => ({ ...acc, [v.uid]: v }), {}) as Record<
  EFilterType,
  FilterOption<any, false>
>;

export const allTypeFilters: FilterObjectClass[] = Object.values(allTypeFieldsData).map(
  (field: FilterOption) =>
    new FilterObjectClass({
      ...field,
      field: field.uid,
      value: undefined,
      // expression: null,
      type: field.type,
    })
);

export const demoFilterString = new FilterObjectClass({
  uid: 'demoFilterString',
  field: demoFieldName.uid,
  label: 'Example String Field',
  value: 'Foo',
  operator: comparisons.contains,
  type: EFilterType.text,
});
export const demoFilterNumber = new FilterObjectClass({
  uid: 'demoFilterNumber',
  field: 'count',
  label: 'Example Number Field',
  value: 0,
  operator: comparisons.greaterThan,
  type: EFilterType.number,
});
// export const demoFilterExpression = new FilterObjectClass({
//   uid: 'demoFilterExpression',
//   field: 'expression',
//   value: 0,
//   operator: '>',
// });

export const demoFilterSelect = new FilterObjectClass({
  uid: 'demoFilterExpression',
  field: demoFieldSelect.uid,
  value: 'a',
  operator: comparisons.contains,
  type: EFilterType.select,
});

export const demoFilterCustom = new FilterObjectClass({
  uid: 'demoFilterCustom',
  field: customField.uid,
  value: 'a',
  operator: comparisons.contains,
  type: 'customFilter',
  isCustomType: true,
});

export const demoFiltersData: FilterObjectClass[] = [
  demoFilterString,
  demoFilterNumber,
  // demoFilterExpression,
  demoFilterSelect,
  demoFilterCustom,
];

export const customFilter = () => false;
export const CustomFilterComponent = ({ filter, updateFilter }) => (
  <div data-testid="customFilter">
    <span>Custom Filter Component</span>

    <button
      onClick={() => {
        const newFilterData = new FilterObjectClass({
          ...filter,
          value: 'button clicked',
        });
        updateFilter(newFilterData);
      }}
    >
      Click me
    </button>
  </div>
);
export const customFilters = { customFilter, customFieldType: customFilter };
export const customFiltersComponents = {
  customFilter: CustomFilterComponent,
  customFieldType: CustomFilterComponent,
};
