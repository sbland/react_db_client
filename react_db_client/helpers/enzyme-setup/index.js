import { configure } from 'enzyme';
import { toMatchDiffSnapshot } from 'snapshot-diff';
import Adapter from 'enzyme-adapter-react-16';


expect.extend({ toMatchDiffSnapshot });
configure({ adapter: new Adapter() });