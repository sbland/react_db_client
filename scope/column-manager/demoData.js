import comparisons from '../../GenericConstants/comparisons';
import filterTypes from '../../GenericConstants/filterTypes';
import FilterObjectClass from '../FilterManager/FilterObjectClass';

export const demoFilterString = new FilterObjectClass({
  uid: 'demoFilterString',
  field: 'name',
  value: 'Foo',
  label: 'Demo Filter String',
  operator: comparisons.contains,
  type: filterTypes.text,
});
export const demoFilterNumber = new FilterObjectClass({
  uid: 'demoFilterNumber',
  field: 'count',
  label: 'Demo Filter Number',
  value: 0,
  operator: comparisons.greaterThan,
  type: filterTypes.number,
});
// export const demoFilterExpression = new FilterObjectClass({
//   uid: 'demoFilterExpression',
//   field: 'expression',
//   label: 'Demo Filter Expression',
//   value: 0,
//   operator: '>',
// });

export const demoFilterSelect = new FilterObjectClass({
  uid: 'demoFilterExpression',
  field: 'select',
  label: 'Demo Filter Select',
  value: 'a',
  operator: comparisons.contains,
  type: filterTypes.select,
});

export const demoFiltersData = [demoFilterString, demoFilterNumber];

export const demoTableDataSimple = [
  {
    uid: 'a',
    name: 'Foo',
    count: 4,
  },
  {
    uid: 'b',
    name: 'Foo',
    count: 3,
  },
  {
    uid: 'c',
    name: 'Car',
    count: -100,
  },
];

// The demo data after filter, sort and evaluate
export const demoTableDataEvaluatedSimple = [
  {
    uid: 'b',
    name: 'Foo',
    count: 3,
    eval: 4,
  },
  {
    uid: 'a',
    name: 'Foo',
    count: 4,
    eval: 5,
  },
];

export const demoHeadingsDataSimple = [
  {
    uid: 'uid',
    label: 'UID',
    type: 'button',
    action: (x) => alert(x),
  },
  {
    uid: 'name',
    label: 'Name',
    type: filterTypes.text,
  },
  {
    uid: 'count',
    label: 'Count',
    type: filterTypes.number,
    defaultValue: 7,
  },
  {
    uid: 'eval',
    label: 'Eval',
    evaluateType: 'number',
    type: filterTypes.number,
    expression: '$count + 1',
    expressionReversed: '$count=$_ - 1',
  },
];

export const demoHeadingsSimpleHiddenIds = demoHeadingsDataSimple
  .filter((h) => h.hidden)
  .map((h) => h.uid);

export const demoDataSimpleTotals = {
  count: 7,
};

export const demoTableData = {
  a: {
    uid: 'a',
    name: 'Foo',
    count: 3,
    multiplier: 2,
    toggle: true,
    readOnly: "Can't touch me",
    select: 'a',
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
  ab: {
    // Note string "4" here to check we can pass string numbers without issues
    uid: 'ab',
    name: 'Foobar',
    count: '99',
    def: 3,
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
  b: {
    uid: 'b',
    name: 'Bar',
    description: 'A really really really long description that needs to fit into a little box!',
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
  c: {
    uid: 'c',
    name: 'C',
    count: 3,
    multiplier: 3,
    expression: 9,
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
  d: {
    uid: 'd',
    name: '',
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
};

export const demoTotals = {
  uid: 'na',
  name: 'na',
  count: 10,
  multiplier: 10,
  toggle: 'na',
  readOnly: 'na',
  select: 'na',
  hiddenDemoNumber: 8,
  hiddenDemo: 'na',
};

export const demoTableDataLong = [...Array(100)]
  .map((v, i) => ({
    uid: i,
    name: 'Foo',
    count: 3,
    multiplier: 2,
    toggle: true,
    readOnly: "Can't touch me",
    select: 'a',
  }))
  .reduce((acc, v) => {
    acc[v.uid] = v;
    return acc;
  }, {});

// export const demoTableData = Object.values(demoTableDataObject);

export const demoTableDataEvaluated = [
  {
    count: 3,
    // def: 3,
    hiddenDemoNumber: 3,
    expression: 6,
    multiplier: 2,
    hiddenDemo: 'Hide me',
    name: 'Foo',
    toggle: true,
    uid: 'a',
    validated: false,
    invalid: 'INVALID',
    invalidb: 'INVALID',
    readOnly: "Can't touch me",
    select: 'a',
  },
  {
    count: 4,
    hiddenDemoNumber: 3,
    def: 3,
    hiddenDemo: 'Hide me',
    expression: 'INVALID',
    invalid: 'INVALID',
    invalidb: 'INVALID',
    name: 'Foobar',
    uid: 'ab',
    validated: 'INVALID',
  },
];

export const demoTableTotals = {
  count: 7,
  def: 3,
  expression: 6,
  hiddenDemoNumber: 6,
  multiplier: 2,
  invalid: 0,
  invalidb: 0,
};

export const demoProcessedData = {
  tableDataFiltered: demoTableDataEvaluated,
  totals: demoTableTotals,
};

export const demoHeadingsData = [
  {
    uid: 'uid',
    label: 'UID',
    type: 'button',
    action: (x) => alert(x),
  },
  {
    uid: 'name',
    label: 'Name',
    type: 'text',
  },
  {
    uid: 'description',
    label: 'description',
    type: 'textLong',
  },
  {
    uid: 'count',
    label: 'Count',
    type: 'number',
    max: 10,
    min: 2,
    def: 8,
    showTotals: true,
  },
  {
    uid: 'def',
    label: 'Def',
    type: 'number',
    max: 10,
    min: 1,
    defaultValue: 7,
    showTotals: true,
  },
  {
    uid: 'multiplier',
    label: 'Multiplier',
    type: 'number',
    max: 10,
    min: 1,
    def: '$def',
  },
  {
    uid: 'readOnly',
    label: 'Read Only',
    type: 'text',
    readOnly: true,
  },
  {
    uid: 'hiddenDemo',
    label: 'Hidden',
    type: 'text',
    hidden: true,
  },
  {
    uid: 'hiddenDemoNumber',
    label: 'Hidden Number',
    type: 'number',
    hidden: true,
  },
  {
    uid: 'toggle',
    label: 'Toggle',
    type: 'toggle',
  },
  {
    uid: 'toggleR',
    label: 'ToggleR',
    type: 'toggle',
    readOnly: true,
  },
  {
    uid: 'expression',
    label: 'Expression',
    evaluateType: 'number',
    type: 'number',
    expression: '$count * $multiplier',
    expressionReversed: '$count=$_ / $multiplier',
    reverseExpression: '$count * $multiplier',
    showTotals: true,
  },
  {
    uid: 'validated',
    label: 'Is Validated',
    type: 'string',
    evaluateType: 'string',
    operator: comparisons.equals,
    target: 'a',
    field: 'select',
    invert: true,
  },
  {
    uid: 'invalid',
    label: 'Invalid',
    evaluateType: 'number',
    type: 'number',
    expression: '$nothing',
  },
  {
    uid: 'invalidb',
    label: 'InvalidB',
    evaluateType: 'number',
    type: 'number',
    expression: '$invalid * 2',
  },
  {
    uid: 'select',
    label: 'Select',
    type: 'select',
    options: [
      { uid: 'a', label: 'A' },
      { uid: 'b', label: 'B' },
      { uid: 'c', label: 'C' },
    ],
  },
  {
    uid: 'internalRoom',
    label: 'Internal Room',
    type: 'entity',
    collection: 'rooms',
  },
  {
    uid: 'longHeading',
    label: 'Long Heading which is really really long',
    type: 'entity',
    collection: 'rooms',
  },
];

// export const DemoTotalsWrapper = () => {
//   const [total, setTotal] = useState(null);
//   return (
//     <div>
//       <DataTable
//         tableData={demoTableData}
//         headingsData={demoHeadingsData}
//         handleSaveData={(data) => console.log(data)}
//         outputTotals={setTotal}
//         showTotals
//       />
//       <h3>Totals</h3>
//       <p className="total">{JSON.stringify(total)}</p>
//     </div>
//   );
// };
