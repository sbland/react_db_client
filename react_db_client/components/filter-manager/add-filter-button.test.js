import '@samnbuk/react_db_client.testing.enzyme-setup';
import React from 'react';
// import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

import {
  comparisons,
  FilterObjectClass,
  filterTypes,
} from '@react_db_client/constants.client-types';

import { AddFilterButton } from './add-filter-button';
import { demoFieldsData } from './demoData';

describe('Add Filter Button Component', () => {
  test('should render', () => {
    shallow(<AddFilterButton fieldsData={demoFieldsData} returnNewFilter={() => {}} />);
  });
  test('should match snapshot', () => {
    const returnNewFilterFn = jest.fn();
    const component = shallow(
      <AddFilterButton fieldsData={demoFieldsData} returnNewFilter={returnNewFilterFn} />
    );
    const tree = component.debug();
    expect(tree).toMatchSnapshot();
  });
  test('should return a new filter when we click add filter btn', () => {
    const returnNewFilterFn = jest.fn();
    const component = mount(
      <AddFilterButton fieldsData={demoFieldsData} returnNewFilter={returnNewFilterFn} />
    );
    const addBtn = component.find('button');
    addBtn.simulate('click');
    expect(returnNewFilterFn).toHaveBeenCalled();
    expect(returnNewFilterFn.mock.calls[0] instanceof FilterObjectClass);
    expect(returnNewFilterFn).toHaveBeenCalledWith(
      new FilterObjectClass({
        field: 'name',
        label: 'Name',
        value: undefined,
        operator: comparisons.equals,
        type: filterTypes.text,
        uid: expect.any(String),
        filterOptionId: 'name',
      })
    );
  });
});
