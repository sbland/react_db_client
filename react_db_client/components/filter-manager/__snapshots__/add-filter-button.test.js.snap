// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Filter Button Component should match snapshot 1`] = `
"<button type=\\"button\\" className=\\"button-two addFilterBtn\\" onClick={[Function: handleNewFilter]}>
  Add Filter
</button>"
`;
